virtcontainers/cgroups.go

```diff
@@ -0,0 +1,111 @@
+// +build linux
+// Copyright (c) 2018 Huawei Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+package virtcontainers
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/kata-containers/runtime/virtcontainers/pkg/annotations"
+	"github.com/opencontainers/runc/libcontainer/cgroups"
+	"github.com/opencontainers/runc/libcontainer/cgroups/fs"
+	"github.com/opencontainers/runc/libcontainer/configs"
+	"github.com/opencontainers/runc/libcontainer/specconv"
+	"github.com/opencontainers/runtime-spec/specs-go"
+)
+
+// cgroupsManager maintains cgroups implementations and config
+type cgroupsManager struct {
+	libcontainerManager cgroups.Manager
+	libcontainerConfig  *configs.Config
+}
+
+// newManager setup cgroup manager for sandbox
+func (cgm *cgroupsManager) newManager(s *Sandbox) error {
+	ociConfigStr, err := s.Annotations(annotations.ConfigJSONKey)
```

Comment:
Where is `s.Annotations()` defined? I can't find it in the PR. Ideally we should only unmarshal the OCI json once for each container, since it is really slow... And for an empty sandbox w/o containers (e.g. in the CRI case), there is no such container OCI spec. You need to handle that case as well. So IMO the cgroup manager needs to be created upon first container creation instead.

Reply:
Yes, because you rely on a container OCI spec to create the cgroup manager and we won't have  a container spec until the first **container** is to be created.

Reply:
Currently the only place unmarshaling the OCI json is in [`kata-agent level`](https://github.com/kata-containers/runtime/blob/c324b55255186d22682bd74cde1beb50c1df8afc/virtcontainers/kata_agent.go#L778), but I think `newManager()` should be called at `sandbox level`.

I would like to:
1. add a pointer named `ociSpec` at `Sandbox struct`;
2. unmarshal in `newSandbox()` and assign it to `ociSpec`;
3. when `createContainer()` in `kata-agent` level, get it from `ociSpec`;

Please share your comments, thanks~

Reply:
@jingxiaolu 

1. Yes, a pointer to the first ocispec in `Sandbox` makes sense since it avoids unmarshalling the same json twice.

2. As I stated above, one issue with `newManager()` in `newSandbox()` is that there may be no containers in sandboxConfig. Then you do not have the OCI spec you need to create the cgroup manager. The right place to do it is in `createContainer()`.

3. Yes, it makes sense.